File: src/dataset/c/binary_search_Exercism_ThomasZumsteg/gemini/GeminiFlash_binary_search_v4.c
#include "binary_search.h"
#include <stddef.h>

int *binary_search(int needle, int *haystack, size_t len) {
    int *low = haystack;
    int *high = haystack + len;

    while (low < high) {
        size_t mid_index = (size_t)(low - haystack) + (size_t)(high - low) / 2;
        int *mid = haystack + mid_index;

        if (*mid < needle) {
            low = mid + 1;
        } else if (*mid > needle) {
            high = mid;
        } else {
            return mid;
        }
    }

    return 0;
}

File: src/dataset/go/binary_search_Exercism_go_ThomasZumsteg/gemini/GeminiFlash_binary_search_v4.go
package binarysearch

import "fmt"

/*SearchInts finds the index of an item in a list using a binary search.*/
func SearchInts(ints []int, find int) int {
	low := 0
	high := len(ints) - 1

	for low <= high {
		mid := low + (high-low)/2 // Prevents potential overflow
		if ints[mid] == find {
			return mid
		} else if ints[mid] < find {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return low
}

/*Message displays the results of a search in a nice way.*/
func Message(ints []int, find int) string {
	n := len(ints)

	if n == 0 {
		return "slice has no values"
	}

	index := SearchInts(ints, find)

	if index < n && ints[index] == find {
		if index == 0 {
			return fmt.Sprintf("%d found at beginning of slice", find)
		} else if index == n-1 {
			return fmt.Sprintf("%d found at end of slice", find)
		} else {
			return fmt.Sprintf("%d found at index %d", find, index)
		}
	} else {
		if index == 0 {
			return fmt.Sprintf("%d < all values", find)
		} else if index == n {
			return fmt.Sprintf("%d > all %d values", find, n)
		} else {
			return fmt.Sprintf("%d > %d at index %d, < %d at index %d",
				find, ints[index-1], index-1, ints[index], index)
		}
	}
}